virtual/activations.go

```diff
@@ -382,6 +383,46 @@ func (a *activations) getServerState() (
 	return a.serverState.serverID, a.serverState.serverVersion
 }
 
+func (a *activations) close(ctx context.Context, numWorkers int) error {
+	a.Lock()
```

Comment:
I see; I did not add this because we use an HTTP server as the "door." So if we want to close the door and not allow any operation, we call it `Shutdown.` The `Shutdown` method waits for all the inflight requests to be made before returning. However, if you still want to add the logic, in case we want something different to an HTTP server in front of the `environment,` I will add that. However, I suggest using atomics instead of locks since they are more performant, and in this case, we do not need complex concurrency access.

Reply:
A really common use case for Nola will be using virtual.Environment embedded as a library so I think it’s important the Environment can manage itself without the HTTP server. In addition when used as a library some of the requests will have originated from the application itself and not from the server so just managing the server doesn’t work

Reply:
I would not use atomics personally. Most times I do it, I regret it. They’re useful for statistics and other things but it’s surprisingly hard to get them right for things like this that “must be right”.

cost of an uncontended RLock is very low, I would not worry about it for now.

Reply:
> I would not use atomics personally. Most times I do it, I regret it. They’re useful for statistics and other things but it’s surprisingly hard to get them right for things like this that “must be right”.

I agree, but that's usually when the atomics are used for complex concurrency patterns. In this case, a flag is straightforward and requires fewer lines than the RLock. However, if it's your taste, then I will change that 👍 NOLA is your project